Don’t force a suspend

swankjesse opened the pull request:
Closes: https://github.com/cashapp/zipline/issues/826

swankjesse committed:
Introduce SuspendResult, to support non-suspending calls

swankjesse committed:
Don't suspend on inbound calls

swankjesse committed:
Restore tracking of serialized services

swankjesse committed:
Get all working tests passing

swankjesse committed:
Get the rest of the tests passing

swankjesse committed:
Make tests more consistent

JakeWharton commented:
Requires more than a quick perusal so I'll look later.

swankjesse committed:
Test the events for the new case

swankjesse committed:
Use a better API to force a suspend

swankjesse commented on zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/calls.kt:
```diff
@@ -296,3 +297,78 @@ internal class ResultSerializer<T>(
+ * Immediate result from invoking a suspending function, that may include either a value (if the
+ * call never suspended) or a cancel callback (if it did suspend).
+ */
+internal class SuspendingResult<T>(
```
Considering for follow-up . . . 

We have a bunch of result types:
 - kotlin.Result - success or failure
 - app.cash.zipline.CallResult - success or failure, encoded JSON, service names
 - app.cash.zipline.internal.bridge.SuspendingResult - success or failure or callback
 - app.cash.zipline.internal.bridge.EncodedSuspendingResult - success or failure or callback, encoded JSON, service names

We also have serializers for kotlin.Result and SuspendingResult.

Some options to reduce how much code we need:
 * Combine the two serializers to save a bit of code. Rename SuspendingResult to InternalResult for symmetry with InternalCall.
 * Change EncodedSuspendingResult to have a `kotlin.Result` as a field

JakeWharton commented on zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/InboundService.kt:
```diff
@@ -68,47 +68,88 @@ internal class InboundService<T : ZiplineService>(
 
-    val cancelCallback = object : CancelCallback, HasPassByReferenceName {
-      override var passbyReferenceName: String? = null
+      deferred.invokeOnCompletion {
```
I'm a bit worried about the Retrofit case here: where `isActive` returns `true` but by the time you get to this code the `async` has completed and thus this lambda will be invoked synchronously (I'm assuming that's the behavior).

It would mean that we invoke the `suspendCallback` below synchronously from the stackframe which invoked the `suspend` function (and before we allow it to return a value). The normal coroutine machinery defends against this (which is what was happening in Retrofit).

In our case, to defend against it, I think we need something like the following:
```kotlin
var scheduleCallback = true
deferred.invokeOnCompletion {
  // ..

  if (scheduleCallback) {
    ziplineDispatcher.schedule({ /* callback logic */ })
  } else {
    // callback logic
  }
}
scheduleCallback = false

swankjesse replied:
It’s single-threaded, so we don‘t need to worry about that.

JakeWharton replied:
As in, the `async` is running on the same dispatcher as this block? That's good.

swankjesse replied:
Exactly.

JakeWharton commented on zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/InboundService.kt:
```diff
@@ -68,47 +68,88 @@ internal class InboundService<T : ZiplineService>(
       }
 
-      override fun toString() = "CancelCallback/$internalCall"
-    }
-    job.invokeOnCompletion {
-      val name = cancelCallback.passbyReferenceName
-      if (name != null) endpoint.remove(name)
-    }
+      val suspendingResult = when {
+        kotlinResult.isFailure -> SuspendingResult<Unit>(failure = kotlinResult.exceptionOrNull()!!)
+        else -> SuspendingResult(success = kotlinResult.getOrNull())
+      }
```
These two `when`'s need merged or something. You create the `kotlin.Result` and then unpack it just to match on it, force unwrap, and then re-wrap it.

swankjesse replied:
Ooooh it’s weirder than that. Added this comment:

```
      // Convert our Deferred<Result<T>> into a Result<T>, which will be failure if _either_ the
      // Deferred failed or if it contains a failure Result.
```

swankjesse replied:
(Is this possible??)

JakeWharton commented on zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/calls.kt:
```diff
@@ -214,6 +214,7 @@ internal class RealCallSerializer(
       return object : SuspendingZiplineFunction<T>(
         name = functionName,
         argSerializers = listOf(),
+        resultSerializer = Int.serializer(), // Placeholder; we're only encoding failures.
```
There's a `NothingSerializer` coming in the next version of kotlinx.coroutines.

swankjesse replied:
Oooh want.

JakeWharton commented on zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/calls.kt:
```diff
@@ -296,3 +297,78 @@ internal class ResultSerializer<T>(
+) : KSerializer<SuspendingResult<T>> {
+
+  override val descriptor: SerialDescriptor = buildClassSerialDescriptor("Result") {
+    element("cancelCallback", ThrowableSerializer.descriptor)
```
```suggestion
    element("cancelCallback", cancelCallbackSerializer.descriptor)
```

JakeWharton commented on zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/calls.kt:
```diff
@@ -296,3 +297,78 @@ internal class ResultSerializer<T>(
+ * call never suspended) or a cancel callback (if it did suspend).
+ */
+internal class SuspendingResult<T>(
+  /** The function suspended. Only non-null for suspend calls. */
+  val cancelCallback: CancelCallback? = null,
+
+  /** The function failed without suspending. */
+  val failure: Throwable? = null,
+
+  /** The function succeeded without suspending. */
+  val success: T? = null,
```
Do you think we need an `init { }` which enforces the invariant that one (and only one) is set? Deserialization below could create an instance with all `null`s, for example.

swankjesse replied:
Good idea. Done.

JakeWharton reviewed (approved):
Cool stuff

swankjesse committed:
Restore a deleted test facet

swankjesse committed:
Update zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/calls.kt

Co-authored-by: Jake Wharton <jw@squareup.com>

swankjesse committed:
Code review feedback

swankjesse commented:
Ugh, I’ve got some unwanted concurrency in the tests.
```
> Process 'Gradle Test Executor 1' finished with non-zero exit value 132
```

swankjesse committed:
Avoid running on multiple threads

swankjesse committed:
Add -Xss2048k to increase the JVM stack size

https://stackoverflow.com/questions/60666890/illegal-instruction-4-from-a-jni-application-on-macos-10-13-6

swankjesse commented:
Okay so eventually I decided to run `java` without Gradle wrapping it, and that was incredibly helpful because I started getting extra information. The JVM printed this before it crashed:

```
Illegal instruction: 4
```

After scouring Stackoverflow for that error I came across [this question](https://stackoverflow.com/questions/60666890/illegal-instruction-4-from-a-jni-application-on-macos-10-13-6), which recommended adding `-Xss2048k` to the Java command. That solved the crashes on x86_64.

It’s Quite Unfortunate that our stacks are deep enough to cause a JVM process crash.

swankjesse committed:
Configure the right stack size
